README.md

# Flutter Float Box

[![alt Become a patron](https://c5.patreon.com/external/logo/become_a_patron_button.png)](https://patreon.com/akshayejh)

Float Box - v1.0

## Overview

[![alt Floating Panel](https://github.com/akshayejh/floating_panel/blob/master/cover.png?raw=true)](https://github.com/akshayejh/floating_box)

Float box is a true floating panel for your app, which can be docked to either edges of the screen (horizontally).

# Code - Basic:

This is what you needs to get the widget working. Import Widget -> Add 'FloatBoxPanel' at the bottom of your stack widget. Add icons inside the "buttons" list.

```
import 'package:widgets/float_box.dart'; // Import Float Box Widget

Stack(
    children: [
        ...

        // Add Float Box Panel at the bottom of the 'stack' widget.
        FloatBoxPanel(
            buttons: [
                // Add Icons to the buttons list.
                Icons.message,
                Icons.photo_camera,
                Icons.video_library
            ]
        );
    ]
);
```

![alt Floating Panel](https://github.com/akshayejh/floating_panel/blob/master/preview_default.gif?raw=true)

# Code - All Properties:

These are all the properties you can customize of the floating panel.


```
    import 'package:widgets/float_box.dart'; // Import Float Box Widget
    
    Stack(
        children: [
            ...
    
            // Add Float Box Panel at the bottom of the 'stack' widget.
            FloatBoxPanel(
                positionTop: 0.0, // Initial Top Position
                positionLeft: 0.0, // Initial Left Position
                backgroundColor: Color(0xFFEDEDED), // Color of the panel
                contentColor: Colors.black, // Color of the icons
                panelShape: PanelShape.rectangle, // 'PanelShape.rectangle' or 'PanelShape.rounded', border radius property doesn't work in the rounded shape
                borderRadius: BorderRadius.circular(8.0), // Border radius of the panel
                dockType: DockType.inside, // 'DockType.inside' or 'DockType.outside', weather to dock the panel outside or inside the edge of the screen
                dockOffset: 5.0, // Offset the dock from the edge depending on the 'dockType' property
                panelAnimDuration: 300, // Duration for panel open and close animation
                panelAnimCurve: Curves.easeOut, // Curve for panel open and close animation
                dockAnimDuration: 300, // Auto dock to the edge of screen - animation duration
                dockAnimCurve: Curves.easeOut, // Auto dock animation curve
                panelOpenOffset: 20.0, // Offset from the edge of screen when panel is open
                panelIcon: Icons.menu, // Panel open/close icon
                size: 70.0, // Size of single button in the panel
                iconSize: 24.0, // Size of icons
                borderWidth: 1.0, // Width of panel border
                borderColor: Colors.black, // Color of panel border
                onPressed: (index) {
                    print("Clicked on item: $index");
                }
                buttons: [
                    // Add Icons to the buttons list.
                    Icons.message,
                    Icons.photo_camera,
                    Icons.video_library
                ]
            );
        ]
    );
```


CHANGELOG.md
## [1.1.3] - Release

* Floating Panel helps you create true floating menu for your app.


test/floatingpanel_test.dart
import 'package:floatingpanel/floatingpanel.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('test', () {
    final floatingPanel = FloatBoxPanel(
      buttons: [
        Icons.message,
        Icons.photo_camera,
        Icons.video_library
      ],
    );
  });
}


example/main.dart
import 'package:floatingpanel/floatingpanel.dart';
import 'package:flutter/material.dart';

class Homepage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Stack(
          children: [
            // Add Float Box Panel at the bottom of the 'stack' widget.
            FloatBoxPanel(
              //Customize properties
              backgroundColor: Color(0xFF222222),
              panelShape: PanelShape.rectangle,
              borderRadius: BorderRadius.circular(8.0),

              buttons: [
                // Add Icons to the buttons list.
                Icons.message,
                Icons.photo_camera,
                Icons.video_library
              ],
            ),
          ],
        ),
      ),
    );
  }
}


lib/floatingpanel.dart
import 'package:flutter/material.dart';

enum PanelShape { rectangle, rounded }

enum DockType { inside, outside }

enum PanelState { open, closed }

class FloatBoxPanel extends StatefulWidget {
  final double positionTop;
  final double positionLeft;
  final Color borderColor;
  final double borderWidth;
  final double size;
  final double iconSize;
  final IconData panelIcon;
  final BorderRadius borderRadius;
  final Color backgroundColor;
  final Color contentColor;
  final PanelShape panelShape;
  final PanelState panelState;
  final double panelOpenOffset;
  final int panelAnimDuration;
  final Curve panelAnimCurve;
  final DockType dockType;
  final double dockOffset;
  final int dockAnimDuration;
  final Curve dockAnimCurve;
  final List<IconData> buttons;
  final Function(int) onPressed;

  FloatBoxPanel(
      {this.buttons,
      this.positionTop,
      this.positionLeft,
      this.borderColor,
      this.borderWidth,
      this.iconSize,
      this.panelIcon,
      this.size,
      this.borderRadius,
      this.panelState,
      this.panelOpenOffset,
      this.panelAnimDuration,
      this.panelAnimCurve,
      this.backgroundColor,
      this.contentColor,
      this.panelShape,
      this.dockType,
      this.dockOffset,
      this.dockAnimCurve,
      this.dockAnimDuration,
      this.onPressed});

  @override
  _FloatBoxState createState() => _FloatBoxState();
}

class _FloatBoxState extends State<FloatBoxPanel> {
  // Required to set the default state to closed when the widget gets initialized;
  PanelState _panelState = PanelState.closed;

  // Default positions for the panel;
  double _positionTop = 0.0;
  double _positionLeft = 0.0;

  // ** PanOffset ** is used to calculate the distance from the edge of the panel
  // to the cursor, to calculate the position when being dragged;
  double _panOffsetTop = 0.0;
  double _panOffsetLeft = 0.0;

  // This is the animation duration for the panel movement, it's required to
  // dynamically change the speed depending on what the panel is being used for.
  // e.g: When panel opened or closed, the position should change in a different
  // speed than when the panel is being dragged;
  int _movementSpeed = 0;

  @override
  void initState() {
    _positionTop = widget.positionTop ?? 0;
    _positionLeft = widget.positionLeft ?? 0;

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // Width and height of page is required for the dragging the panel;
    double _pageWidth = MediaQuery.of(context).size.width;
    double _pageHeight = MediaQuery.of(context).size.height;

    // All Buttons;
    List<IconData> _buttons = widget.buttons;

    // Dock offset creates the boundary for the page depending on the DockType;
    double _dockOffset = widget.dockOffset ?? 20.0;

    // Widget size if the width of the panel;
    double _widgetSize = widget.size ?? 70.0;

    // **** METHODS ****

    // Dock boundary is calculated according to the dock offset and dock type.
    double _dockBoundary() {
      if (widget.dockType != null && widget.dockType == DockType.inside) {
        // If it's an 'inside' type dock, dock offset will remain the same;
        return _dockOffset;
      } else {
        // If it's an 'outside' type dock, dock offset will be inverted, hence
        // negative value;
        return -_dockOffset;
      }
    }

    // If panel shape is set to rectangle, the border radius will be set to custom
    // border radius property of the WIDGET, else it will be set to the size of
    // widget to make all corners rounded.
    BorderRadius _borderRadius() {
      if (widget.panelShape != null &&
          widget.panelShape == PanelShape.rectangle) {
        // If panel shape is 'rectangle', border radius can be set to custom or 0;
        return widget.borderRadius ?? BorderRadius.circular(0);
      } else {
        // If panel shape is 'rounded', border radius will be the size of widget
        // to make it rounded;
        return BorderRadius.circular(_widgetSize);
      }
    }

    // Total buttons are required to calculate the height of the panel;
    double _totalButtons() {
      if (widget.buttons == null) {
        return 0;
      } else {
        return widget.buttons.length.toDouble();
      }
    }

    // Height of the panel according to the panel state;
    double _panelHeight() {
      if (_panelState == PanelState.open) {
        // Panel height will be in multiple of total buttons, I have added "1"
        // digit height for each button to fix the overflow issue. Don't know
        // what's causing this, but adding "1" fixed the problem for now.
        return (_widgetSize + (_widgetSize + 1) * _totalButtons()) +
            (widget.borderWidth ?? 0);
      } else {
        return _widgetSize + (widget.borderWidth ?? 0) * 2;
      }
    }

    // Panel top needs to be recalculated while opening the panel, to make sure
    // the height doesn't exceed the bottom of the page;
    void _calcPanelTop() {
      if (_positionTop + _panelHeight() > _pageHeight + _dockBoundary()) {
        _positionTop = _pageHeight - _panelHeight() + _dockBoundary();
      }
    }

    // Dock Left position when open;
    double _openDockLeft() {
      if (_positionLeft < (_pageWidth / 2)) {
        // If panel is docked to the left;
        return widget.panelOpenOffset ?? 30.0;
      } else {
        // If panel is docked to the right;
        return ((_pageWidth - _widgetSize)) - (widget.panelOpenOffset ?? 30.0);
      }
    }

    // Panel border is only enabled if the border width is greater than 0;
    Border _panelBorder() {
      if (widget.borderWidth != null && widget.borderWidth > 0) {
        return Border.all(
          color: widget.borderColor ?? Color(0xFF333333),
          width: widget.borderWidth ?? 0.0,
        );
      } else {
        return null;
      }
    }

    // Force dock will dock the panel to it's nearest edge of the screen;
    void _forceDock() {
      // Calculate the center of the panel;
      double center = _positionLeft + (_widgetSize / 2);

      // Set movement speed to the custom duration property or '300' default;
      _movementSpeed = widget.dockAnimDuration ?? 300;

      // Check if the position of center of the panel is less than half of the
      // page;
      if (center < _pageWidth / 2) {
        // Dock to the left edge;
        _positionLeft = 0.0 + _dockBoundary();
      } else {
        // Dock to the right edge;
        _positionLeft = (_pageWidth - _widgetSize) - _dockBoundary();
      }
    }

    // Animated positioned widget can be moved to any part of the screen with
    // animation;
    return AnimatedPositioned(
      duration: Duration(
        milliseconds: _movementSpeed,
      ),
      top: _positionTop,
      left: _positionLeft,
      curve: widget.dockAnimCurve ?? Curves.fastLinearToSlowEaseIn,

      // Animated Container is used for easier animation of container height;
      child: AnimatedContainer(
        duration: Duration(milliseconds: widget.panelAnimDuration ?? 600),
        width: _widgetSize,
        height: _panelHeight(),
        decoration: BoxDecoration(
          color: widget.backgroundColor ?? Color(0xFF333333),
          borderRadius: _borderRadius(),
          border: _panelBorder(),
        ),
        curve: widget.panelAnimCurve ?? Curves.fastLinearToSlowEaseIn,
        child: Wrap(
          direction: Axis.horizontal,
          children: [
            // Gesture detector is required to detect the tap and drag on the panel;
            GestureDetector(
              onPanEnd: (event) {
                setState(
                  () {
                    _forceDock();
                  },
                );
              },
              onPanStart: (event) {
                // Detect the offset between the top and left side of the panel and
                // x and y position of the touch(click) event;
                _panOffsetTop = event.globalPosition.dy - _positionTop;
                _panOffsetLeft = event.globalPosition.dx - _positionLeft;
              },
              onPanUpdate: (event) {
                setState(
                  () {
                    // Close Panel if opened;
                    _panelState = PanelState.closed;

                    // Reset Movement Speed;
                    _movementSpeed = 0;

                    // Calculate the top position of the panel according to pan;
                    _positionTop = event.globalPosition.dy - _panOffsetTop;

                    // Check if the top position is exceeding the dock boundaries;
                    if (_positionTop < 0 + _dockBoundary()) {
                      _positionTop = 0 + _dockBoundary();
                    }
                    if (_positionTop >
                        (_pageHeight - _panelHeight()) - _dockBoundary()) {
                      _positionTop =
                          (_pageHeight - _panelHeight()) - _dockBoundary();
                    }

                    // Calculate the Left position of the panel according to pan;
                    _positionLeft = event.globalPosition.dx - _panOffsetLeft;

                    // Check if the left position is exceeding the dock boundaries;
                    if (_positionLeft < 0 + _dockBoundary()) {
                      _positionLeft = 0 + _dockBoundary();
                    }
                    if (_positionLeft >
                        (_pageWidth - _widgetSize) - _dockBoundary()) {
                      _positionLeft =
                          (_pageWidth - _widgetSize) - _dockBoundary();
                    }
                  },
                );
              },
              onTap: () {
                setState(
                  () {
                    // Set the animation speed to custom duration;
                    _movementSpeed = widget.panelAnimDuration ?? 200;

                    if (_panelState == PanelState.open) {
                      // If panel state is "open", set it to "closed";
                      _panelState = PanelState.closed;

                      // Reset panel position, dock it to nearest edge;
                      _forceDock();

                      print("Float panel closed.");
                    } else {
                      // If panel state is "closed", set it to "open";
                      _panelState = PanelState.open;

                      // Set the left side position;
                      _positionLeft = _openDockLeft();

                      _calcPanelTop();

                      print("Float Panel Open.");
                    }
                  },
                );
              },
              child: _FloatButton(
                size: widget.size ?? 70.0,
                icon: widget.panelIcon ?? Icons.add,
                color: widget.contentColor ?? Colors.white,
                iconSize: widget.iconSize ?? 24.0,
              ),
            ),
            Visibility(
              visible: true, // TODO:: _contentVisibility,
              child: Container(
                child: Column(
                  children: List.generate(_buttons.length, (index) {
                    return GestureDetector(
                      onTap: () {
                        widget.onPressed(index);
                      },
                      child: _FloatButton(
                        size: widget.size ?? 70.0,
                        icon: _buttons[index] ?? Icons.add,
                        color: widget.contentColor ?? Colors.white,
                        iconSize: widget.iconSize ?? 24.0,
                      ),
                    );
                  }),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _FloatButton extends StatelessWidget {
  final double size;
  final Color color;
  final IconData icon;
  final double iconSize;

  _FloatButton({this.size, this.color, this.icon, this.iconSize});

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.white.withOpacity(0.0),
      width: size ?? 70.0,
      height: size ?? 70.0,
      child: Icon(
        icon ?? Icons.add,
        color: color ?? Colors.white,
        size: iconSize ?? 24.0,
      ),
    );
  }
}
